tools/tpm2_readpublic.c

```diff
@@ -57,6 +58,13 @@ static tool_rc read_public_and_save(ESYS_CONTEXT *ectx) {
         }
     }
 
+    if (ctx.public_path)
```

Comment:
The `TPM2B_PUBLIC` is saved via the call to `tpm2_convert_pubkey_save()` with the format of `pubkey_format_tss`, which contains the `TPMT_PUBLIC` data, whats the reason that doesn't work for you?

Reply:
+1 to what Bill said. You can specify this in the command line option with the option: 
--format tss

Reply:
i have split in two commits. first gives a bit of rational in the commit message.

Point is that TPM2B_PUBLIC is the TSS format. And that's great if you want to feed it back into TPM tools.

That said, we face the challenge that we need to calculate the "Name" we got from a tpm2_certify from a public area dump someone exported with tpm2-tools.

According to "Architecture Spec", a "Name" is built like: nameAlg || hash(TPMT_PUBLIC), so the TPM2B_PUBLIC won't work without further processing. Having a TPMT_PUBLIC export (which --public is about) would be convenient and also feels like it fits the spirit of this command (e.g. called readpublic).

For the name algorithm, check chapter 16 for persistent objects here: https://trustedcomputinggroup.org/wp-content/uploads/TCG_TPM2_r1p55_Part1_Architecture_pubrev.pdf

Let me know your thoughts.

Reply:
> i have split in two commits. first gives a bit of rational in the commit message.
> 
> Point is that TPM2B_PUBLIC is the TSS format. And that's great if you want to feed it back into TPM tools.

The default format is a TPM2B_PUBLIC which is a size followed by a TPMT_PUBLIC, so you can just go past the size field and get the TPMT_PUBLIC structure. 

You could just use DD, a bit "hacky" but it would work.
```sh
tpm2_readpublic -c primary.ctx -o tpm2b_public.dat

# skip the UINT16 size field.
dd if=tpm2b_public.dat of=tpmt_public.dat skip=2 bs=1

cat tpmt_public.dat | openssl sha256
(stdin)= 914a1abc3f6b71166e4b8d3a02ee338524bb5b82a7c9e1a59f28dc97f7f2fd0c
```
Add the `00` and the algorithm `0b` and you get the full name:
`000b914a1abc3f6b71166e4b8d3a02ee338524bb5b82a7c9e1a59f28dc97f7f2fd0c`

> 
> That said, we face the challenge that we need to calculate the "Name" we got from a tpm2_certify from a public area dump someone exported with tpm2-tools.

If you need the name, it's provided by the YAML output dumped to stdout.

```sh
$ ./tools/tpm2_readpublic -c object.ctx
name: 000b914a1abc3f6b71166e4b8d3a02ee338524bb5b82a7c9e1a59f28dc97f7f2fd0c
qualified name: 000bdb4f165a0db93bc1cccbc5266f69ceaafe1bf96bcab11ca73d6f155a15a66591
name-alg:
  value: sha256
  raw: 0xb
```
> 
> According to "Architecture Spec", a "Name" is built like: nameAlg || hash(TPMT_PUBLIC), so the TPM2B_PUBLIC won't work without further processing. Having a TPMT_PUBLIC export (which --public is about) would be convenient and also feels like it fits the spirit of this command (e.g. called readpublic).

If you want the TPMT_PUBLIC I am not opposed to doing it, just opposed to your implementation. Their is already a -o for output and -f for format. So it would be adding a
format selector, something like tss-no-size so it saves the first structure embedded in the TPM2B_, in this case the TPMT.

Adding another option isn't the right approach.

> 
> For the name algorithm, check chapter 16 for persistent objects here: https://trustedcomputinggroup.org/wp-content/uploads/TCG_TPM2_r1p55_Part1_Architecture_pubrev.pdf
> 
> Let me know your thoughts.

Reply:
Actually for the format string name, just use tpmt if you choose to add a format specified.